### PR TITLE
storage: add WITH (DISK) support to upsert sources

### PR DESCRIPTION
**Pull request by guswynn**
Part of https://github.com/MaterializeInc/materialize/issues/17067

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

- The top commit is just basic tests. @philip-stoev my plan to eventually move all upsert-related tests into this new mzcompose, and test both backends there. I wanted to keep this pr manageable. I also added this new composition to the tests ci pipeline.
- The bottom commit is the meat of the implementation; It contains:
  -  Support for `WITH (ON DISK)` in `CREATE SOURCE` (note that `ON DISK = true/false` are explicitly banned for clarity, and simplifying migrations in the future.
  - `--instance-directory` is the new optional plan required to use this feature. This is forward to `clusterd` from `envd`. Naming (and the naming of `InstanceContext`) isn't great, and I'm not attached to anything specific here)
  - The rest of the changes are mostly boilerplate piping things through. Note that error handling is lackluster, but thats a separate task in the epic. The non `ON DISK` flow is left unchanged.
  - Note that this is behind `--unsafe-mode`. Next step is to change this to an LD flag. This will happen in a followup pr, to avoid too much noise in one pr.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
   (see notion)
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:394]: Agree we should try to find a better name than this! It is EC2 specific in its current incarnation, and "instance" is already a massively overloaded term. Perhaps `ephemeral_directory` or `scratch_directory` or somesuch.

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:655]: Validation of Kubernetes orchestrator-specific parameters should happen in the Kubernetes orchestrator constructor, not here.

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:739]: This validation should happen inside of environmentd/lib.rs, not here, since it's not related to the CLI arguments specifically.

File: src/storage-client/src/types/sources.rs
Comment (benesch) [inline src/storage-client/src/types/sources.rs:1746]: It'd be great if you could do a pass for missing docstrings before putting the PR up for review!

File: src/sql/src/plan/with_options.rs
Comment (benesch) [inline src/sql/src/plan/with_options.rs:255]: I would rather not add this. This is complexity that we will likely forget to remove. It also is likely that someone will use this in a non-unsafe option in the future, at which point it becomes product debt, because we will have some special boolean options that do not support the consistent `OPTION = {true|false}` syntax. The fact that all boolean options support that syntax is a feature!

File: src/sql/src/plan/statement/ddl.rs
Comment (benesch) [inline src/sql/src/plan/statement/ddl.rs:4158]: Copy pasta.

File: src/storage/src/server.rs
Comment (benesch) [inline src/storage/src/server.rs:38]: This is "instance" as in "storage instance" right?

```suggestion
    /// Other configuration for storage instances.
```

File: src/sql-parser/src/parser.rs
Comment (benesch) [inline src/sql-parser/src/parser.rs:2589]: I'm sorry for not realizing this when I made the original suggestion, but I'm wondering if we should make the option `DISK [= true]` or `SCRATCH DISK [= true]`. Just in case we need to stabilize the option. WDYT? Does that read better to y'all than "on disk"? It does to my ear, but maybe that' sjust me.

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [reply src/environmentd/src/bin/environmentd/main.rs:655]: Hmm, actually, I think we should not do this validation at all! I can't see why it would need to be absolute for the Kubernetes orchestrator. It needs to be the _same_ for all spawned services, sure, but a relative path presumably works just fine.

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [reply src/environmentd/src/bin/environmentd/main.rs:739]: Wait, actually, I don't think envd should be creating the scratch directory at all! It should be the individual clusters.

File: src/controller/src/clusters.rs
Comment (benesch) [inline src/controller/src/clusters.rs:460]: This doesn't seem like it will work correctly for the process orchestrator—they'll all end up sharing the same instance directory, no?

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [reply src/environmentd/src/bin/environmentd/main.rs:394]: Actually, I'm pretty sure this needs to be process orchestrator specific. The Kubernetes orchestrator will conjure up a scratch directory upon request, by creating a volume mount of the right type.

**Review by benesch — COMMENTED** [review]
Left some comments within! I think something is funny about how this is currently plumbed. I think the orchestrators need to become aware of scratch directory-enabled services. In particular, I think `ServiceConfig` should learn a `scratch_dir: Option<usize>` parameter that specifies how big of a scratch directory to create. The process orchestrator will create this scratch directory in the argument specified by `--orchestrator-process-scratch-dir` (which will be required when the scratch dir is in use), and the Kubernetes orchestrator will create this scratch directory by mounting a volume of the right type (which will have to wait until cloud tells us what that is). Then the type of `ServiceConfig::args` should evolve from

```
&'a (dyn Fn(&BTreeMap<String, String>) -> Vec<String> + Send + Sync)
```

to

```
struct ServiceAssignment<'a> {
    ports: &'a BTreeMap<String, String>,
    scratch_dir: Option<String>,
}

&'a (dyn Fn(&ServiceAssignment) -> Vec<String> + Send + Sync)
```

and from that the controller can generate the correct `--scratch-dir` argument for the clusterd process.

**Comment by philip-stoev** [comment]
@guswynn As a first step, I am going to manually patch a couple of CI jobs so that the new option is exercised in a few scenarios. Once the final format of the SQL syntax and the command-line option is finalized we can look at how to test this systematically. I may need an option that enables this across the board, not per-source, so that the `CREATE SOURCE` directives from the existing tests do not have to be modified.

File: test/upsert/kafka-upsert-debezium-sources.td
Comment (philip-stoev) [inline test/upsert/kafka-upsert-debezium-sources.td:101]: Unfortunately this approach of patching the `WITH` clause is good for a single smoketest testdrive test but does not scale. A global setting will be needed that simply enables the feature across the board for all `CREATE SOURCE` statements.

File: src/storage-client/src/types/sources.rs
Comment (guswynn) [reply src/storage-client/src/types/sources.rs:1746]: I usually do but I missed this one

File: src/sql-parser/src/parser.rs
Comment (guswynn) [reply src/sql-parser/src/parser.rs:2589]: `DISK` seems good to me!

File: test/upsert/kafka-upsert-debezium-sources.td
Comment (guswynn) [reply test/upsert/kafka-upsert-debezium-sources.td:101]: Is this something we can do in the testdrive binary itself?

Comment (guswynn) [reply test/upsert/kafka-upsert-debezium-sources.td:101]: How can `testdrive` discover all the create source statements and do this? Does it parse the sql?

**Comment by philip-stoev** [comment]
Item No 1:

```
thread 'thread 'timely:work-1timely:work-2' panicked at '' panicked at 'thread 'instance directory to be there if rendering an ON DISK sourceinstance directory to be there if rendering an ON DISK source', timely:work-0', /var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0629d720fba03823c-1/materialize/tests/src/storage/src/render/upsert.rs' panicked at '/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0629d720fba03823c-1/materialize/tests/src/storage/src/render/upsert.rs:instance directory to be there if rendering an ON DISK source:157', :157/var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0629d720fba03823c-1/materialize/tests/src/storage/src/render/upsert.rs14:
```

This apparently means that the directory does not exist?

File: src/storage-client/src/types/sources.rs
Comment (guswynn) [reply src/storage-client/src/types/sources.rs:1746]: Added a pre-push git hook locally

**Comment by philip-stoev** [comment]
ITem No 2. I see rocksdb files being created, but they never grow and are not updated, even though the sources have hundreds of thousands of records:

````
materialize@2a38dab7817d:/mzdata/rocksdb$ du --max-depth=1 -h
196K	./u45
196K	./u41
196K	./u33
196K	./u8
196K	./u4
196K	./u13
196K	./u29
196K	./u21
196K	./u17
196K	./u37
```

I see the computed process having e.g. the `LOG` files  open, but I do not think there are any writes coming to those files. @guswynn can you check the disk usage on the Mz instance that you use for development?

**Comment by philip-stoev** [comment]
Item No 3. DROP SOURCE does not seem to clean up the respective rocksdb database.

File: src/controller/src/clusters.rs
Comment (philip-stoev) [reply src/controller/src/clusters.rs:460]: Within the instance directory, separate subdirectories are being created for each source -- `u1234`, `u2345` and so on. So it seems that multiple computeds can successfully coexist inside the same container

File: src/sql/src/plan/with_options.rs
Comment (guswynn) [reply src/sql/src/plan/with_options.rs:255]: sounds good!

**Comment by guswynn** [comment]
@aljoscha this should be good to go!

To address @philip-stoev's concerns:

- Item number 1: I believe we resolved this offline and is unblocked
- Item number 2: I believe we resolved this offline. Data is not persisted until it has to be.
- Item number 3: is fixed in the latest commit (it only works for `IN CLUSTER` sources, and has some subtleties. @aljoscha we may have to improve this in the future)

Also, I added 2 LD flags: one that sets the default (this simplifies the testdrive test I added), and also one that that allows to use of `DISK` (as opposed to requiring `unsafe` mode)

For @benesch's points, all I believe are addressed. This behavior is currently now process-orchestrator only (zippy tests will need to wait)

**Comment by guswynn** [comment]
@philip-stoev This is mostly to unblock a basic working way to use the rocksdb backing, so i can rebase https://github.com/MaterializeInc/materialize/pull/18912, so is it fine if we merge and iterate on the testing?

**Comment by philip-stoev** [comment]
Yes, sorry for the delay, I spend like 2 days fighting a stupid thing in mzcompose.

I enabled  this by default for the entire CI in a separate branch and both the per-push CI and Nightly are green. I have also started a Release Qualification Test.

**Review by philip-stoev — APPROVED** [review]

**Comment by philip-stoev** [comment]
@guswynn can you do a git commit --amend push and force it,  just so that the CI runs again? For some reason the last run did not complete.

File: src/clusterd/src/bin/clusterd.rs
Comment (benesch) [inline src/clusterd/src/bin/clusterd.rs:166]:
```suggestion
    /// A scratch directory that can be used for ephemeral storage.
```

File: src/environmentd/src/bin/environmentd/main.rs
Comment (benesch) [inline src/environmentd/src/bin/environmentd/main.rs:369]:
```suggestion
    /// A scratch directory that orchestrated processes can use for ephemeral storage.
```

File: src/orchestrator-process/src/lib.rs
Comment (benesch) [inline src/orchestrator-process/src/lib.rs:153]:
```suggestion
    /// A scratch directory that orchestrated processes can use for ephemeral storage.
```

**Review by benesch — COMMENTED** [review]
Orchestrator changes LGTM, with some docco nits.

File: src/sql-parser/tests/testdata/ddl
Comment (jkosh44) [inline src/sql-parser/tests/testdata/ddl:1706]: Can you add a test that tries to give `DISK` a value to make sure it returns an error.

File: src/sql/src/plan/statement/ddl.rs
Comment (jkosh44) [reply src/sql/src/plan/statement/ddl.rs:948]: I think `bail_unsupported` is fine, it provides a more structured error than `sql_bail`.

File: src/sql/src/session/vars.rs
Comment (jkosh44) [inline src/sql/src/session/vars.rs:533]: This should be named `enable_upsert_source_disk` to match the other feature flag naming conventions.

File: src/sql/src/session/vars.rs
Comment (jkosh44) [inline src/sql/src/session/vars.rs:536]: The description should formatted as "Feature flag indicating whether ... (Materialize)." to match the other feature flags.

**Review by jkosh44 — COMMENTED** [review]
Couple small nits on the surfaces side.

File: src/sql-parser/tests/testdata/ddl
Comment (guswynn) [reply src/sql-parser/tests/testdata/ddl:1706]: It actually works with `= true` now, so ill try that!

**Comment by guswynn** [comment]
@benesch has approved the orchestrator changes
@jkosh44 has signed off on the surfaces changes
I checked with @aljoscha, and after fixing nits, the storage changes are approved
@philip-stoev after ci passes I will merge this as there is no outstanding review or issues, as far as I can tell

File: src/sql/src/session/vars.rs
Comment (jkosh44) [inline src/sql/src/session/vars.rs:537]: Very small nit, but the period should go after `(Materialize)`.

File: src/sql/src/session/vars.rs
Comment (jkosh44) [inline src/sql/src/session/vars.rs:1785]: The name and comment for this function still need to change.

**Pull request merged**
